### PR TITLE
Use Set hash in Distinct (2nd attempt)

### PR DESCRIPTION
**Pull request by dashpole**
Re-opening https://github.com/open-telemetry/opentelemetry-go/pull/5028 after new benchmarks.

```
goos: linux
goarch: amd64
pkg: go.opentelemetry.io/otel/attribute
cpu: Intel(R) Xeon(R) CPU @ 2.20GHz
                                            │   main.txt    │              hash.txt               │
                                            │    sec/op     │    sec/op     vs base               │
EquivalentMapAccess/Empty-24                    32.01n ± 2%   10.12n ±  4%  -68.37% (p=0.002 n=6)
EquivalentMapAccess/1_string_attribute-24      106.25n ± 2%   10.01n ±  5%  -90.58% (p=0.002 n=6)
EquivalentMapAccess/10_string_attributes-24   826.250n ± 1%   9.982n ± 11%  -98.79% (p=0.002 n=6)
EquivalentMapAccess/1_int_attribute-24         106.65n ± 2%   10.13n ±  3%  -90.50% (p=0.002 n=6)
EquivalentMapAccess/10_int_attributes-24       833.25n ± 2%   10.04n ±  5%  -98.80% (p=0.002 n=6)
geomean                                         190.3n        10.06n        -94.72%
```

```
goos: linux
goarch: amd64
pkg: go.opentelemetry.io/otel/sdk/metric
cpu: Intel(R) Xeon(R) CPU @ 2.20GHz
                                                         │   main.txt    │              hash.txt               │
                                                         │    sec/op     │    sec/op     vs base               │
SyncMeasure/NoView/Int64Counter/Attributes/0-24             264.8n ± 12%   316.1n ± 30%  +19.37% (p=0.002 n=6)
SyncMeasure/NoView/Int64Counter/Attributes/1-24             370.9n ± 25%   303.3n ± 18%  -18.24% (p=0.002 n=6)
SyncMeasure/NoView/Int64Counter/Attributes/10-24           2038.5n ± 15%   320.6n ± 17%  -84.27% (p=0.002 n=6)
SyncMeasure/NoView/Float64Counter/Attributes/0-24           237.3n ±  9%   279.5n ± 21%  +17.78% (p=0.015 n=6)
SyncMeasure/NoView/Float64Counter/Attributes/1-24           418.9n ± 17%   294.9n ± 18%  -29.61% (p=0.002 n=6)
SyncMeasure/NoView/Float64Counter/Attributes/10-24         1753.5n ±  3%   308.1n ± 13%  -82.43% (p=0.002 n=6)
SyncMeasure/NoView/Int64UpDownCounter/Attributes/0-24       277.9n ± 15%   280.7n ± 10%        ~ (p=0.937 n=6)
SyncMeasure/NoView/Int64UpDownCounter/Attributes/1-24       368.5n ± 25%   275.8n ±  7%  -25.13% (p=0.002 n=6)
SyncMeasure/NoView/Int64UpDownCounter/Attributes/10-24     1762.0n ± 14%   274.1n ± 10%  -84.45% (p=0.002 n=6)
SyncMeasure/NoView/Float64UpDownCounter/Attributes/0-24     262.1n ± 11%   281.4n ± 12%        ~ (p=0.180 n=6)
SyncMeasure/NoView/Float64UpDownCounter/Attributes/1-24     369.1n ± 14%   281.8n ± 12%  -23.68% (p=0.002 n=6)
SyncMeasure/NoView/Float64UpDownCounter/Attributes/10-24   1757.5n ± 37%   294.5n ±  8%  -83.24% (p=0.002 n=6)
SyncMeasure/NoView/Int64Histogram/Attributes/0-24           308.3n ± 10%   322.1n ± 12%        ~ (p=0.818 n=6)
SyncMeasure/NoView/Int64Histogram/Attributes/1-24           307.2n ± 12%   334.0n ± 18%        ~ (p=0.240 n=6)
SyncMeasure/NoView/Int64Histogram/Attributes/10-24         1137.0n ± 18%   359.0n ± 10%  -68.43% (p=0.002 n=6)
SyncMeasure/NoView/Float64Histogram/Attributes/0-24         316.4n ± 11%   313.6n ± 11%        ~ (p=0.818 n=6)
SyncMeasure/NoView/Float64Histogram/Attributes/1-24         354.0n ± 15%   349.6n ±  6%        ~ (p=0.699 n=6)
SyncMeasure/NoView/Float64Histogram/Attributes/10-24        941.1n ± 20%   366.2n ± 11%  -61.08% (p=0.002 n=6)
geomean                                                     533.3n         307.4n        -42.36%
```

I think we should strongly consider adopting this, or something similar. In particular, it has a very large performance impact when more than a few attributes are used.

### CODE REVIEW

File: attribute/set.go
Comment (pellared) [inline attribute/set.go:1]: Shouldn't the following part of `Set` GoDoC comment be removed?

```go
	// This type will remain comparable for backwards compatibility. The
	// equivalence of Sets across versions is not guaranteed to be stable.
	// Prior versions may find two Sets to be equal or not when compared
	// directly (i.e. ==), but subsequent versions may not. Users should use
	// the Equals method to ensure stable equivalence checking.
	//
	// Users should also use the Distinct returned from Equivalent as a map key
	// instead of a Set directly. In addition to that type providing guarantees
	// on stable equivalence, it may also provide performance improvements.
```

File: attribute/set.go [hunk 1]
```diff
@@ -46,15 +49,22 @@ type (
 	Sortable []KeyValue
 )
 
+// Compile time check these types remain comparable.
+var (
+	_ = isComparable(Set{})
```
Comment (pellared) [inline attribute/set.go:53]: "tracers", "aggregations", "loggers" currently are stored in`map[instrumentation.Scope]XYZ`. However, `instrumentation.Scope` has `attribute.Set` field. This means this PR would be these maps would be "safe to use" as currently they should use `attribute.Distinct`. This is for me another argument for this change.

File: attribute/set.go
Comment (MrAlias) [reply attribute/set.go:1]: These statements still are still true. Why remove it?

File: attribute/internal/fnv/fnv.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+	return h.Uint64(math.Float64bits(val))
+}
+
+func (h Hash) Int64(val int64) Hash {
+	return h.Uint64(uint64(val)) // nolint:gosec // overflow doesn't matter since we are hashing.
+}
+
+func (h Hash) String(val string) Hash {
+	v := uint64(h)
+	for _, c := range val {
```
Comment (MrAlias) [inline attribute/internal/fnv/fnv.go]: This iterates on runes not bytes like the FNV-1a algorithm is designed (and the built-in one does).

```suggestion
	for i := 0; i < len(val); i++ {
```

Non-ASCII test cases are needed to catch this bug.

File: attribute/hash.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+		for i := 0; i < rv.Len(); i++ {
+			h = h.String(rv.Index(i).String())
+		}
+	case INVALID:
+	default:
+		// Logging is an alternative, but using the internal logger here
+		// causes an import cycle so it is not done.
+		v := kv.Value.AsInterface()
+		msg := fmt.Sprintf("unknown value type: %[1]v (%[1]T)", v)
+		panic(msg)
```
Comment (MrAlias) [inline attribute/hash.go:89]: This was a useful default for a prototype. Not sure if we want this in the actual release. Though, I'm not sure what it should be replaced with.

File: attribute/hash_test.go
Comment (MrAlias) [inline attribute/hash_test.go:1]: We likely needs some fuzzing tests here.

File: attribute/set.go [hunk 4]
```diff
@@ -383,7 +391,7 @@ func computeDistinctReflect(kvs []KeyValue) any {
 
 // MarshalJSON returns the JSON encoding of the Set.
 func (l *Set) MarshalJSON() ([]byte, error) {
```
Comment (MrAlias) [inline attribute/set.go:407]: Not sure if JSON unmarshal already works, but it is definitely not going to work with the addition of the hash. We need to verify if this is breaking that ability.

**Review by MrAlias — COMMENTED** [review]
We need to consider what is an acceptable level of hash collision here. We may want to use a larger bit size for the hash or look into ways to avoid the collision if we determine the collision rate too high. We may also consider this approach acceptable.

File: attribute/set.go
Comment (pellared) [reply attribute/set.go:1]: You are right. My bad :wink:

**Comment by MrAlias** [comment]
For reference, Prometheus has used as similar size hash for their implementation. There haven't been any known issues related to collisions there and it has been out for ~ 10 years.

File: attribute/hash.go (hunk 3 above)
Comment (MrAlias) [reply attribute/hash.go:89]: Leave the panic.

File: attribute/set.go (hunk 4 above)
Comment (MrAlias) [reply attribute/set.go:407]: We should add tests to cover this.

**Comment by juliusv** [comment]
Some input from the sidelines (and yay on the performance improvements! 🥳):

> For reference, Prometheus has used as similar size hash for their implementation. There haven't been any known issues related to collisions there and it has been out for ~ 10 years.

It looks like collision handling was added to `client_golang` around a decade ago: https://github.com/prometheus/client_golang/pull/221

Besides the client library, we have seen some real collisions over the years in the Prometheus server that have all been addressed:

* Over a decade ago, a server with millions of series at SoundCloud ran into hash collisions. I think we used an FNV-1a hash back then to differentiate series within the TSDB. All this has completely changed by now, so the TSDB is resistant against this now.
* Another real hash collision happened more recently in the PromQL engine in Prometheus: https://github.com/prometheus/prometheus/issues/5724. But as far as I can tell, this was due to a bad custom way of how we hashed labels together.

We still use hashes without collision checking in some places, which are hopefully unlikely enough to ever cause a collision. See also https://github.com/prometheus/prometheus/issues/12519 for more discussion on hash collision potential in Prometheus.

File: attribute/hash_test.go [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+	attrs := make([]KeyValue, len(keyVals))
+	for i := range keyVals {
+		attrs[i] = keyVals[i]("k")
+	}
+
+	var h fnv.Hash
+
+	b.ResetTimer()
+	b.ReportAllocs()
+	for n := 0; n < b.N; n++ {
```
Comment (bwplotka) [inline attribute/hash_test.go]: You could simplify (same in other benchmarks):

```suggestion
	for range b.N {
```

I assume we can't use `b.Loop()` until 1.23 support is dropped.

File: attribute/hash_test.go [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+		h = hashKVs(attrs)
```
Comment (bwplotka) [inline attribute/hash_test.go]: I believe `_ = h` is no much better than `_ = hashKVs(attrs)` as  `_ = h` is likely removed by a compiler. More safer approach would be `var Sink fnv.Hash` global, but generally it's quite hard (and clear as you will see 0ms) to get our benchmarked code optimized away; also https://go.dev/blog/testing-b-loop helps.

I would simply do `_ = hashKVs` here.

Comment (dashpole) [reply attribute/hash_test.go]: I did see a difference when I made `h` global here, so I did that and removed the empty assignment below. https://github.com/open-telemetry/opentelemetry-go/pull/7175/commits/162f87dfe02e7e2d5477f141530e162e17d05cf5

File: attribute/hash_test.go (hunk 5 above)
Comment (dashpole) [reply attribute/hash_test.go]: https://github.com/open-telemetry/opentelemetry-go/pull/7175/commits/95f1bedbd273ebfcad087624004358be418e12ba

File: attribute/internal/fnv/fnv.go (hunk 2 above)
Comment (dashpole) [reply attribute/internal/fnv/fnv.go]: Fixed and added a test.  I verified that the test fails without the fix. https://github.com/open-telemetry/opentelemetry-go/pull/7175/commits/0516d7192b21cfe35629a324c603e6a5e2be306d

File: CHANGELOG.md [hunk 7]
```diff
@@ -73,6 +74,7 @@ The next release will require at least [Go 1.24].
 - Fix `go.opentelemetry.io/otel/exporters/prometheus` to deduplicate suffixes if already present in metric name when UTF8 is enabled. (#7088)
 - Fix the `go.opentelemetry.io/otel/exporters/stdout/stdouttrace` self-observability component type and name. (#7195)
 - Fix partial export count metric in `go.opentelemetry.io/otel/exporters/stdout/stdouttrace`. (#7199)
+- Clarify the documentation about equivalence guarantees for the `Set` and `Distinct` types in `go.opentelemetry.io/otel/attribute`. (#7175)
```
Comment (pellared) [inline CHANGELOG.md]: Sorry if this is a dumb question (maybe I am too tired after checking a dozen of PRs today).
Where is the clarification in documentation?

Comment (dashpole) [reply CHANGELOG.md]: I copied that from the original PR :).  I'll update it.

**Comment by MrAlias** [comment]
@dashpole are we going to look into collision remediation techniques based on the Prometheus groups experience?

**Comment by dashpole** [comment]
Yes, I will look into what was done in https://github.com/prometheus/client_golang/pull/221.

**Comment by dashpole** [comment]
The TL;DR is that handling collisions just means using a `map[uint64][]metricWithLabelValues` instead of a `map[uint64]metricWithLabelValues` to store aggregations (note the slice), and iterating over elements to find the exact one. Looks fairly straightforward. See https://github.com/prometheus/client_golang/blob/84a4734fe73aecc9ec36e1fbcb09c9a2d679f392/prometheus/vec.go#L319

I think the biggest difference for the attributes package is that we would need to ensure that `Equals()` correctly handles hash collisions (i.e. it should check more than that the set's Distinct are the same).  Alternatively, we could switch to the "Add a new Hash() function" approach since that keeps a correct `Equals()` implementation. I'm not sure if the current Distinct implementation has poor `==` performance or not...

File: CHANGELOG.md (hunk 7 above)
Comment (dashpole) [reply CHANGELOG.md]: removed.

File: attribute/hash_test.go
Comment (dashpole) [reply attribute/hash_test.go:1]: added a fuzz test

File: attribute/set.go (hunk 4 above)
Comment (dashpole) [reply attribute/set.go:407]: https://github.com/open-telemetry/opentelemetry-go/pull/7268

Comment (dashpole) [reply attribute/set.go:407]: The test passes on this PR, so we should be all good.

**Comment by dashpole** [comment]
There is a significant performance cost (the cost of set.Equals from https://github.com/open-telemetry/opentelemetry-go/pull/7262) to check for duplicates, but it is still a huge improvement over the current state.  I'm trying to add a unit test with a collision, but it is taking a while for the fuzzer to generate a collision for me... Even with a cardinality of 100 million attributes, I didn't hit one :).  I updated the benchmark results in the description.

**Comment by dashpole** [comment]
Looking a little bit more at hash collision probabilities, i'm not sure it is worth implementing hash collision resolution.

Unlike the prometheus server, the metric name isn't a label, so the name doesn't count towards cardinality for the purposes of hash collision.  So for us to reach 1 million cardinality in a `map` in the SDK would require 1 million series _per instrument_.  Even at 1 million series on a single instrument, the collision probability should be 2.5x10^-8 if hashes are random, which is quite a low chance.  For a "normal" 1000 cardinality on an instrument, it is ~2.7x10^-17...

The overhead of collision detection is high.  For the `SyncMeasure/NoView/Int64Counter/Attributes/10` case, it more than doubles the performance cost (104 ns -> 225 ns) of Measure to add collision detection.  This is from the Equals() method.

**Comment by juliusv** [comment]
> Unlike the prometheus server, the metric name isn't a label, so the name doesn't count towards cardinality for the purposes of hash collision. So for us to reach 1 million cardinality in a `map` in the SDK would require 1 million series _per instrument_.

In the Prometheus client library (vs. the server you mentioned), the collision handling is also happening on a per metric name / instrument basis. But from what I can tell from reading https://github.com/prometheus/client_golang/pull/221, the reason for introducing collision handling within each instrument was not due to actual observations of collisions, but due to a vague general feeling of unease about potential collisions. Probably it's actually fine to not do anything.

**Comment by dashpole** [comment]
I've removed collision handling based on the above, and re-ran benchmarks in the description.  The PoC is here for future reference: https://github.com/open-telemetry/opentelemetry-go/commit/cff8970eab6f2c72b1fa73803558ba3511abdf7c

I believe this is ready to go!

**Review by pellared — APPROVED** [review]
:+1: 

It would be great to have @MrAlias review/approval given his initial work: https://github.com/open-telemetry/opentelemetry-go/pull/5028

File: attribute/hash_test.go [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+		kvs := []KeyValue{
+			String(k1, s),
+			Int(k2, i),
+			Int64(k3, i64),
+			Float64(k4, f),
+			Bool(k5, b),
+		}
+		h = hashKVs(kvs)
+	})
+}
```
Comment (MrAlias) [inline attribute/hash_test.go]: Adding fuzz testing is a great addition. There are a few limitations of this current test though:

1. Limited Type Coverage: Only tests 5 basic types (string, int, int64, float64, bool) but misses all slice types
2. Fixed Structure: Always creates exactly 5 KeyValues, doesn't test variable-length slices
3. No Edge Cases: Doesn't test special values like NaN, infinity, empty strings, empty slices
4. No Collision Detection: Doesn't verify that different inputs produce different hashes
5. No Key Collision Testing: Doesn't test duplicate keys or key variations
6. Limited String Testing: Only tests ASCII strings, not Unicode/UTF-8 edge cases

This could be improved. For example:

```suggestion
func FuzzHashKVs(f *testing.F) {
	// Add seed inputs to ensure coverage of edge cases.
	f.Add("", "", "", "", "", "", 0, int64(0), 0.0, false, uint8(0))
	f.Add("key", "value", "🌍", "test", "bool", "float", -1, int64(-1), -1.0, true, uint8(1))
	f.Add("duplicate", "duplicate", "duplicate", "duplicate", "duplicate", "NaN", 0, int64(0), math.Inf(1), false, uint8(2))
	
	f.Fuzz(func(t *testing.T, k1, k2, k3, k4, k5, s string, i int, i64 int64, fVal float64, b bool, sliceType uint8) {
		// Test variable number of attributes (0-10).
		numAttrs := len(k1) % 11 // Use key length to determine number of attributes.
		if numAttrs == 0 && len(k1) == 0 {
			// Test empty set.
			h := hashKVs(nil)
			if h == 0 {
				t.Error("hash of empty slice should not be zero")
			}
			return
		}
		
		var kvs []KeyValue
		
		// Add basic types.
		if numAttrs > 0 {
			kvs = append(kvs, String(k1, s))
		}
		if numAttrs > 1 {
			kvs = append(kvs, Int(k2, i))
		}
		if numAttrs > 2 {
			kvs = append(kvs, Int64(k3, i64))
		}
		if numAttrs > 3 {
			kvs = append(kvs, Float64(k4, fVal))
		}
		if numAttrs > 4 {
			kvs = append(kvs, Bool(k5, b))
		}
		
		// Add slice types based on sliceType parameter
		if numAttrs > 5 {
			switch sliceType % 4 {
			case 0:
				// Test BoolSlice with variable length.
				bools := make([]bool, len(s)%5) // 0-4 elements
				for i := range bools {
					bools[i] = (i+len(k1))%2 == 0
				}
				kvs = append(kvs, BoolSlice("boolslice", bools))
			case 1:
				// Test IntSlice with variable length.
				ints := make([]int, len(s)%6) // 0-5 elements  
				for i := range ints {
					ints[i] = i + len(k2)
				}
				kvs = append(kvs, IntSlice("intslice", ints))
			case 2:
				// Test Int64Slice with variable length.
				int64s := make([]int64, len(s)%4) // 0-3 elements
				for i := range int64s {
					int64s[i] = int64(i) + i64
				}
				kvs = append(kvs, Int64Slice("int64slice", int64s))
			case 3:
				// Test Float64Slice with variable length and special values.
				float64s := make([]float64, len(s)%5) // 0-4 elements
				for i := range float64s {
					switch i % 4 {
					case 0:
						float64s[i] = fVal
					case 1:
						float64s[i] = math.Inf(1) // +Inf
					case 2:
						float64s[i] = math.Inf(-1) // -Inf  
					case 3:
						float64s[i] = math.NaN() // NaN
					}
				}
				kvs = append(kvs, Float64Slice("float64slice", float64s))
			}
		}
		
		// Add StringSlice.
		if numAttrs > 6 {
			strings := make([]string, len(k1)%4) // 0-3 elements
			for i := range strings {
				strings[i] = fmt.Sprintf("%s_%d", s, i)
			}
			kvs = append(kvs, StringSlice("stringslice", strings))
		}
		
		// Test duplicate keys (should be handled by Set construction).
		if numAttrs > 7 && len(k1) > 0 {
			kvs = append(kvs, String(k1, "duplicate_key_value"))
		}
		
		// Add more attributes with Unicode keys.
		if numAttrs > 8 {
			kvs = append(kvs, String("🔑", "unicode_key"))
		}
		if numAttrs > 9 {
			kvs = append(kvs, String("empty", ""))
		}
		
		// Sort to ensure consistent ordering (as Set would do).
		slices.SortFunc(kvs, func(a, b KeyValue) int {
			return cmp.Compare(string(a.Key), string(b.Key))
		})
		
		// Remove duplicates (as Set will do).
		if len(kvs) > 1 {
			j := 0
			for i := 1; i < len(kvs); i++ {
				if kvs[j].Key != kvs[i].Key {
					j++
					kvs[j] = kvs[i]
				} else {
					// Keep the later value for duplicate keys.
					kvs[j] = kvs[i]
				}
			}
			kvs = kvs[:j+1]
		}
		
		// Hash the key-value pairs.
		h1 := hashKVs(kvs)
		h2 := hashKVs(kvs) // Should be deterministic
		
		if h1 != h2 {
			t.Errorf("hash is not deterministic: %d != %d for kvs=%v", h1, h2, kvs)
		}
		
		if h1 == 0 && len(kvs) > 0 {
			t.Errorf("hash should not be zero for non-empty input: kvs=%v", kvs)
		}
		
		// Test that different inputs produce different hashes (most of the time).
		// This is a probabilistic test - collisions are possible but rare.
		if len(kvs) > 0 {
			// Modify one value slightly.
			modifiedKvs := make([]KeyValue, len(kvs))
			copy(modifiedKvs, kvs)
			if len(modifiedKvs) > 0 {
				switch modifiedKvs[0].Value.Type() {
				case STRING:
					modifiedKvs[0] = String(string(modifiedKvs[0].Key), modifiedKvs[0].Value.AsString()+"_modified")
				case INT64:
					modifiedKvs[0] = Int64(string(modifiedKvs[0].Key), modifiedKvs[0].Value.AsInt64()+1)
				case BOOL:
					modifiedKvs[0] = Bool(string(modifiedKvs[0].Key), !modifiedKvs[0].Value.AsBool())
				case FLOAT64:
					val := modifiedKvs[0].Value.AsFloat64()
					if !math.IsNaN(val) && !math.IsInf(val, 0) {
						modifiedKvs[0] = Float64(string(modifiedKvs[0].Key), val+1.0)
					}
				}
				
				h3 := hashKVs(modifiedKvs)
				// Note: We don't assert h1 != h3 because hash collisions are theoretically possible
				// but we can log suspicious cases for manual review.
				if h1 == h3 && !reflect.DeepEqual(kvs, modifiedKvs) {
					t.Logf("Potential hash collision detected: original=%v, modified=%v, hash=%d", kvs, modifiedKvs, h1)
				}
			}
		}
	})
}
```

Comment (pellared) [reply attribute/hash_test.go]: Why

```
		// Store result to prevent optimization.
		h = h1
```

?

File: attribute/hash_test.go [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+func FuzzHashKVs(f *testing.F) {
+	f.Fuzz(func(_ *testing.T, k1, k2, k3, k4, k5, s string, i int, i64 int64, f float64, b bool) {
+		kvs := []KeyValue{
+			String(k1, s),
+			Int(k2, i),
+			Int64(k3, i64),
+			Float64(k4, f),
+			Bool(k5, b),
+		}
+		h = hashKVs(kvs)
```
Comment (pellared) [inline attribute/hash_test.go]: this is not a benchmark

```suggestion
		hashKVs(kvs)
```

File: attribute/hash_test.go [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+	attrs := make([]KeyValue, len(keyVals))
+	for i := range keyVals {
+		attrs[i] = keyVals[i]("k")
+	}
+
+	b.ResetTimer()
+	b.ReportAllocs()
+	for range b.N {
+		h = hashKVs(attrs)
+	}
```
Comment (pellared) [inline attribute/hash_test.go]:
```suggestion
	for b.Loop() {
		hashKVs(attrs)
	}
```

File: attribute/hash_test.go (hunk 8 above)
Comment (MrAlias) [reply attribute/hash_test.go]: Ah, good catch. That's from prior iterations.

Updating.

File: attribute/set.go [hunk 11]
```diff
@@ -144,20 +151,29 @@ func (l *Set) ToSlice() []KeyValue {
...
-	return l.Equivalent() == o.Equivalent()
+	if l.Equivalent() != o.Equivalent() {
+		return false
+	}
+	if l == nil || l.hash == 0 {
+		l = emptySet
+	}
+	if o == nil || o.hash == 0 {
+		o = emptySet
+	}
```
Comment (MrAlias) [inline attribute/set.go]: This has a ... frustrating bug. The empty set is actually mutable because it is not a set but a file-level pointer exposed to the user:

```go
func TestSetEqualsEmpty(t *testing.T) {
	e := attribute.EmptySet()
	empty := *e

	alt := attribute.NewSet(attribute.String("A", "B"))
	*e = alt

	var s attribute.Set
	assert.Truef(t, s.Equals(&empty), "expected %v to equal empty set %v", s, attribute.EmptySet())
}
```

```console
> go test -run=TestSetEqualsEmpty
--- FAIL: TestSetEqualsEmpty (0.00s)
    set_test.go:547:
        	Error Trace:	/home/tyler/go/src/go.opentelemetry.io/otel/attribute/set_test.go:547
        	Error:      	Should be true
        	Test:       	TestSetEqualsEmpty
        	Messages:   	expected {0 <nil>} to equal empty set &{3226491390042623715 [{A {4 0 B <nil>}}]}
FAIL
```

To ensure this is not allowed, we likely need to define a new `zeroSet` (or some other name) that is not exposed to the user and will avoid this type of bug.

Comment (MrAlias) [reply attribute/set.go]: Maybe rename `emptySet` to `userDefinedEmpty` and create a new `emptySet` that isn't a pointer?

**Comment by bboreham** [comment]
> For reference, Prometheus has used as similar size hash for their implementation. There haven't been any known issues related to collisions there and it has been out for ~ 10 years.

Julius already cited this, but I too want to note there were issues:
* https://github.com/prometheus/prometheus/issues/5724

Prometheus has generally moved from `fnv` to `xxhash` which is much better behaved.

* Citation for `xxhash` being better than `fnv`: https://github.com/prometheus/client_golang/issues/633#issuecomment-521200201 (unfortunately there is no evidence there, just a statement).
* The first place I can see `xxhash` added (2016): https://github.com/prometheus/prometheus/commit/c6cd9989058ad6562c5c4d83e012c622da8e45ff
* A utility to deliberately search for collisions: https://github.com/pstibrany/labels_hash_collisions

File: attribute/set.go (hunk 11 above)
Comment (MrAlias) [reply attribute/set.go]: https://github.com/open-telemetry/opentelemetry-go/issues/7356

**Comment by dashpole** [comment]
I plan to investigate xxHash in a separate issue. From my brief research, it is primarily designed for efficient hashing of large amounts of input data (e.g. checksums for files).  It should still be faster than fnv64-1a for our inputs, but the difference may or may not be significant.  I've seen conflicting statements for which algorithm has a better distribution of hash values, although most seem to favor xxHash as well. I don't think that should block this PR considering the substantial performance benefits it brings.

**Review by MrAlias — APPROVED** [review]
